test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+      asyncTestLastCheck = Date.now();
+      if (arguments.length > 0) return fail(err);
+      setTimeout(nextAsyncTest, 10);
+    }
+
+    if (asyncTestQueue.length) {
+      var test = asyncTestQueue.shift();
+      currentTest = test;
+      test.action(done);
+    } else {
+      clearInterval(asyncTestHandle);
+    }
+  }
+
+  return function asyncTest(fn) {
+    var stack = new Error().stack.split('\n').slice(1).join('\n');
+    asyncTestQueue.push({action: fn, stack: stack});
+    if (!asyncTestsEnabled) {
+      asyncTestsEnabled = true;
+      asyncTestLastCheck = Date.now();
+      process.on('uncaughtException', fail);
+      asyncTestHandle = setInterval(function() {
+        var now = Date.now();
+        if (now - asyncTestLastCheck > 10000) {
+          return fail(new Error('Async test timeout exceeded'));
+        }
+      }, 10);
+      setTimeout(nextAsyncTest, 10);
+    }
+  };
+
+})();
+
+function setupException(fn) {
+  var listeners = process.listeners('uncaughtException');
+  process.removeAllListeners('uncaughtException');
+  process.on('uncaughtException', fn);
+  return function clean() {
+    process.removeListener('uncaughtException', fn);
+    listeners.forEach(function(listener) {
+      process.on('uncaughtException', listener);
+    });
+  };
+}
+
+function clean() {
+  process.removeAllListeners('unhandledRejection');
+  process.removeAllListeners('rejectionHandled');
+}
+
+function onUnhandledSucceed(done, predicate) {
+  clean();
+  process.on('unhandledRejection', function(reason, promise) {
+    try {
+      predicate(reason, promise);
+    } catch (e) {
+      return done(e);
+    }
+    done();
+  });
+}
+
+function onUnhandledFail(done) {
+  clean();
+  process.on('unhandledRejection', function(reason, promise) {
+    done(new Error('unhandledRejection not supposed to be triggered'));
+  });
+  process.on('rejectionHandled', function() {
+    done(new Error('rejectionHandled not supposed to be triggered'));
+  });
+  setTimeout(function() {
+    done();
+  }, 10);
+}
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  Promise.reject(e);
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    reject(e);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    setImmediate(function() {
+      reject(e);
+    });
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+  });
+  new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+    }, 1);
+  });
+});
+
+asyncTest(function(done) {
```

Comment:
Catching a promise rejection after setImmediate is not soon enough to stop unhandledRejection